File: src/main/kotlin/org/araqnid/typevariance/kotlin/EventHandler.kt
package org.araqnid.typevariance.kotlin

import org.araqnid.typevariance.BasicEvent
import org.araqnid.typevariance.Event
import java.time.Instant

interface EventHandler<in E> {
    fun accept(event: E, timestamp: Instant)
}

val eventHandler: EventHandler<BasicEvent> = object : EventHandler<Event> {
    override fun accept(event: Event, timestamp: Instant) = TODO()
}

val basicEventHandler: EventHandler<BasicEvent> = object : EventHandler<BasicEvent> {
    override fun accept(event: BasicEvent, timestamp: Instant) = TODO()
}

// val invalid: EventHandler<Event> = basicEventHandler

inline fun <reified E : Event> downcast(underlying : EventHandler<E>) : EventHandler<Event> {
    return object : EventHandler<Event> {
        override fun accept(event: Event, timestamp: Instant) {
            underlying.accept(event as E, timestamp)
        }
    }
}

val runtimeSafety: EventHandler<Event> = downcast(basicEventHandler)

File: src/main/scala/org/araqnid/typevariance/scala/EventHandler.scala
package org.araqnid.typevariance.scala

import java.time.Instant

import org.araqnid.typevariance.{BasicEvent, Event}

import scala.reflect.ClassTag

trait EventHandler[-E] {
  def accept(event: E, timestamp: Instant)
}

object EventHandlers {
  val eventHandler: EventHandler[BasicEvent] = new EventHandler[Event] {
    override def accept(event: Event, timestamp: Instant): Unit = ???
  }

  val basicEventHandler: EventHandler[BasicEvent] = new EventHandler[BasicEvent] {
    override def accept(event: BasicEvent, timestamp: Instant): Unit = ???
  }

  // val invalid: EventHandler[Event] = basicEventHandler

  def downcast[E <: Event : ClassTag](underlying: EventHandler[E]): EventHandler[Event] = {
    new EventHandler[Event] {
      override def accept(event: Event, timestamp: Instant): Unit = {
        underlying.accept(implicitly[ClassTag[E]].runtimeClass.cast(event).asInstanceOf[E], timestamp)
      }
    }
  }

  val runtimeSafety: EventHandler[Event] = downcast(basicEventHandler)
}
